controllers/containerizedworkload_controller.go

```diff
@@ -0,0 +1,156 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"github.com/pkg/errors"
+	"time"
+
+	cpv1alpha1 "github.com/crossplaneio/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/go-logr/logr"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	oamv1alpha2 "github.com/oam-dev/core-resource-controller/api/v1alpha2"
+)
+
+const (
+	oamReconcileWait = 30 * time.Second
+)
+
+// Reconcile error strings.
+const (
+	errRenderWorkload  = "cannot render workload"
+	errUpdateStatus    = "cannot apply status"
+	errApplyDeployment = "cannot apply the deployment"
+	errApplyService    = "cannot apply the service"
+	errGCDeployment    = "cannot clean up stale deployments"
+	errScaleDeployment = "cannot scale the deployment"
+)
+
+// ContainerizedWorkloadReconciler reconciles a ContainerizedWorkload object
+type ContainerizedWorkloadReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=core.oam.dev,resources=containerizedworkloads,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core.oam.dev,resources=containerizedworkloads/status,verbs=get;update;patch
+func (r *ContainerizedWorkloadReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	log := r.Log.WithValues("containerizedworkload", req.NamespacedName)
+	log.Info("Reconcile container workload")
+
+	var workload oamv1alpha2.ContainerizedWorkload
+	if err := r.Get(ctx, req.NamespacedName, &workload); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	log.Info("Get the workload")
+
+	deploy, err := r.renderWorkload(ctx, &workload)
+	if err != nil {
+		workload.Status.SetConditions(cpv1alpha1.ReconcileError(errors.Wrap(err, errRenderWorkload)))
+		log.Error(err, "Failed to render a deployment")
+		return reconcile.Result{RequeueAfter: oamReconcileWait}, errors.Wrap(r.Status().Update(ctx, &workload),
+			errUpdateStatus)
+	}
+	log.Info("Successfully rendered a deployment", "deployment", deploy.Name)
+
+	// server side apply, only the fields we set are touched
+	applyOpts := []client.PatchOption{client.ForceOwnership, client.FieldOwner(workload.ObjectMeta.Name)}
+	if err := r.Patch(ctx, deploy, client.Apply, applyOpts...); err != nil {
+		workload.Status.SetConditions(cpv1alpha1.ReconcileError(errors.Wrap(err, errApplyDeployment)))
+		log.Error(err, "Failed to apply to a deployment")
+		return reconcile.Result{RequeueAfter: oamReconcileWait}, errors.Wrap(r.Status().Update(ctx, &workload),
+			errUpdateStatus)
+	}
+	log.Info("Successfully applied a deployment", "UID", deploy.UID)
```

Comment:
Server side apply is a pretty new feature - I believe only Kubernetes 1.16 and up? https://godoc.org/github.com/crossplane/crossplane-runtime/pkg/resource#Apply has similar functionality but is compatible with older API servers. It will only set the fields we changed, and leave any other fields alone.

Reply:
I believe that  I looked but not sure I understand the subtle difference between them. I will look more

Reply:
I can say that `resource.Apply`:

1. Stores a deepcopy of the supplied object as the desired state.
1. If the supplied object doesn't exist in the API server, it's created and the function returns.
1. If the supplied object does exist, reads its state from the API server into the supplied object.
1. Optionally, ensures the latest object has the same controller reference as the desired object.
1. Overwrites any fields of the latest object with the fields of the desired object.

I believe server side apply does effectively the same thing in a more advanced and probably better fashion, but `resource.Apply` has worked well enough for us and works with older API servers.